net/portmapper/probe.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package portmapper
+
+import (
+	"context"
+	"time"
+
+	"tailscale.com/syncs"
+)
+
+// Prober periodically pings the network and checks for port-mapping services.
+type Prober struct {
+	// stop will stop the prober
+	stop func()
+
+	// Each of the SubResults below is intended to expose whether a specific service is available
+	// for use on a client, and the most recent seen time. Should not be modified externally, and
+	// will be periodically updated.
+
+	// PMP stores the result of probing pmp services and is populated by the prober.
+	PMP syncs.WaitableResult
+	// PCP stores the result of probing pcp services and is populated by the prober.
+	PCP syncs.WaitableResult
+	// UPnP stores the result of probing pcp services and is populated by the prober.
+	UPnP syncs.WaitableResult
+}
+
+// newProber will start a prober if it does not exist on the given portmapping client.
+// Should be called when the client's lock is held.
+func (c *Client) newProber(ctx context.Context) {
+	stop := false
+	p := &Prober{
+		PMP:  syncs.NewWaitableResult(),
+		PCP:  syncs.NewWaitableResult(),
+		UPnP: syncs.NewWaitableResult(),
+		stop: func() { stop = true },
+	}
+	c.prober = p
+	go func() {
+		for {
+			res, err := c.oldProbe(ctx)
+			p.PMP.Set(res.PMP, err)
+			p.PCP.Set(res.PCP, err)
+			p.UPnP.Set(res.UPnP, err)
+
+			time.Sleep(trustServiceStillAvailableDuration * 3 / 4)
+			if stop {
+				return
+			}
+		}
+	}()
+}
+
+// Stop gracefully turns the Prober off, completing the current probes before exiting.
+//
+// Calling stop Close multiple times will have no additional effects.
+func (p *Prober) Close() { p.stop() }
+
+// CurrentStatus returns the current results of the prober, regardless of whether they have
```

Comment:
what's the error value on incomplete? it's not clear how callers should handle its errors. is it an error to be incomplete? is it an error to not have 1 or more services?